src/pip/_internal/models/link.py

```diff
@@ -300,6 +308,12 @@ def from_json(
         if metadata_info is None:
             metadata_info = file_data.get("dist-info-metadata")
 
+        upload_time: datetime.datetime | None
```

Comment:
```suggestion
```

mypy is smart enough to infer optional types if None is assigned in an else branch.